src/Semantic/Graph.hs

```diff
@@ -71,7 +71,29 @@ import           Text.Show.Pretty (ppShow)
 
 data GraphType = ImportGraph | CallGraph
 
-type AnalysisClasses = '[ Declarations1, Eq1, Evaluatable, FreeVariables1, AccessControls1, Foldable, Functor, Ord1, Show1 ]
+-- | Constraints we require for a term’s syntax in order to analyze it.
+class
+  ( Declarations1 syntax
+  , Eq1 syntax
+  , Evaluatable syntax
+  , FreeVariables1 syntax
+  , AccessControls1 syntax
+  , Foldable syntax
+  , Functor syntax
+  , Ord1 syntax
+  , Show1 syntax
+  ) => AnalysisClasses syntax
+instance
+  ( Declarations1 syntax
+  , Eq1 syntax
+  , Evaluatable syntax
+  , FreeVariables1 syntax
+  , AccessControls1 syntax
+  , Foldable syntax
+  , Functor syntax
+  , Ord1 syntax
+  , Show1 syntax
+  ) => AnalysisClasses syntax
```

Comment:
I’m not a huge fan of this, but it allows us to cut the number of generated coercions by 4/5ths. On the other hand, that doesn’t seem to do a lot to compile times, at least at `-O1`.

We can’t use the abstracted parsing facilities for this module, because far too much of the machinery is coupled to `Term` to be able to usefully abstract it. `Evaluatable` itself takes a constructor parameterized by subterms, which won’t work for precise ASTs. But that’s alright, because we wouldn’t want to anyway—the whole idea is for precise ASTs to be compiled to core and analyzed there instead.

(The same goes for the `Semantic.AST` stuff, which I’d honestly rather we just got rid of outright.)